polyply/src/apply_links.py

```diff
@@ -364,8 +396,12 @@ def run_molecule(self, meta_molecule):
         """
         molecule = meta_molecule.molecule
         force_field = meta_molecule.force_field
+        resnames = set(nx.get_node_attributes(molecule, "resname").values())
 
         for link in tqdm(force_field.links):
+            link_resnames = _get_link_resnames(link)
+            if not _resnames_match(resnames, link_resnames):
+                continue
```

Comment:
The point really is to only check against single matches. This is an initial step of filtering to not do expensive graph matching and atom reconstruction on Links that for sure do not apply. Usually you cannot check that all resnames match, because links apply to multiple residues and combinations thereof. They are most of the time indiscriminately set by the resnames flag in the molecule directive. I've tested this a lot and this is a safe way (i.e. it can always be tested no matter how links are written) for speeding up when you have large libraries of diverse links.

Reply:
FYI later on a more detailed resname check is made, so this is really just speedup.

Reply:
Sure, but you may be able to speed it up (significantly) more by making the preliminary check (this one) a little bit more expensive, but much more restrictive. Up to you, would require a bit of profiling anyway.